Redirect "ch_" charge id format to "pi_" payment intent id on Transaction Detail screens

EduardoUmpierre opened the pull request:
Fixes #3122

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

Redirect "ch_" charge id format to "pi_" payment intent id on Transaction Detail screens

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

Update the Transaction Details screen to allow receiving a "pi_" Payment Intent ID instead of a "ch_" Charge ID. This change should create more consistency between the links and the URLs on the client.

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Perform an order on your test shop
* Go to **WooCommerce -> Orders**
* Open on the order that you have made
* Click on the Payment Intent id link 
![image](https://user-images.githubusercontent.com/16882226/173832383-39fedf12-39e6-451f-9990-33467c1de14e.png)
* It should have "ch_" on the query param and then redirect to the "pi_" equivalent

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

luizreis commented on client/payment-details/test/index.js:
```diff
@@ -0,0 +1,168 @@
+	)
+);
+
+global.wcSettings = {
+	countries: {
+		US: 'United States of America',
+	},
+};
+
+global.wcpaySettings = {
+	featureFlags: {
+		paymentTimeline: true,
+	},
+	zeroDecimalCurrencies: [ 'usd' ],
+	connect: {
+		country: 'US',
+		availableCountries: {
+			US: 'United States (US)',
+		},
+	},
+	currencyData: {
+		US: {
+			code: 'USD',
+			symbol: '$',
+			symbolPosition: 'left',
+			thousandSeparator: ',',
+			decimalSeparator: '.',
+			precision: 2,
+		},
+	},
+};
```
Sorry if this is a stupid question, but why do we need to set all of this here?

EduardoUmpierre replied:
No, it's not a stupid question! I have added all this data while testing, but in fact, those aren't really necessary. All the unnecessary data was removed on [105d935](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/105d935a0904f88f8bb92ca485fdeafd5d75a108). Thanks for raising it! 👍

luizreis commented on client/payment-details/test/index.js:
```diff
@@ -0,0 +1,168 @@
+	},
+};
+
+describe( 'Payment details page', () => {
```
Kudos on covering the component redirection with some tests! 🚀

luizreis commented on client/payment-details/test/index.js:
```diff
@@ -0,0 +1,168 @@
+			<PaymentDetailsPage query={ chargeQuery } />
+		);
+
+		expect( container ).toMatchSnapshot();
```
I checked this snapshot and it doesn't seem to be using the charge data in the amounts. Is that expected?

EduardoUmpierre replied:
Yes, because it represents the loading state from `charge_id` to `payment_intent_id` query param. We can confirm that by seeing the `is-loadable-placeholder` class that is added to the elements. I added a snapshot test to confirm the behavior when passing a `payment_intent_id` as query param. Thanks for raising that!

EduardoUmpierre commented on client/data/payment-intents/reducer.js:
```diff
@@ -0,0 +1,32 @@
+ */
+import TYPES from './action-types';
+
+const receivePaymentIntent = ( state = {}, { type, id, data, error } ) => {
```
Sure, it makes sense to be `receivePaymentIntents`. Fixed on [4a5c13e](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/4a5c13e9942f6e9b7bb310e2e12dad40b0d0cd06). Thanks for raising this!

EduardoUmpierre commented on client/data/payment-intents/test/resolvers.js:
```diff
@@ -0,0 +1,85 @@
+
+const errorResponse = { code: 'error' };
+
+const chargeResponse = {
```
Yes, it was a left-over from another file 😳. Fixed on [3b06ee7](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/3b06ee7a7bcb9504cd21e1ca25435ded1a33793a).

EduardoUmpierre commented on client/data/payment-intents/test/resolvers.js:
```diff
@@ -0,0 +1,85 @@
+	} );
+
+	describe( 'on error', () => {
+		test( 'should update state with error on error', () => {
```
Fixed on [5b8a374](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/5b8a374f4c2cb8f9eaf9b161c2500c49c0dc45ad).

EduardoUmpierre commented on client/data/payment-intents/hooks.js:
```diff
@@ -0,0 +1,79 @@
+
+			const paymentIntent = getPaymentIntent( id );
+
+			if ( paymentIntent?.charge?.id ) {
+				const { id: chargeId } = paymentIntent.charge;
+
+				return getChargeData( chargeId, selectors );
+			}
```
We currently don't receive the same charge object from the payment intent request as we get from the charge request. It will require changes on the server to retrieve the expanded data. This work will be part of [Use payment intents in payment details view](https://github.com/Automattic/woocommerce-payments/issues/353#).

EduardoUmpierre commented on client/data/payment-intents/index.js:
```diff
@@ -0,0 +1,12 @@
+import * as resolvers from './resolvers';
+
+export { reducer, selectors, actions, resolvers };
+export * from './hooks';
```
Yes, I agree we should use Typescript on new files. Fixed on [04b6286](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/04b628653b113539691dd4677a4703a02f87f1ef).

EduardoUmpierre commented on client/data/timeline/hooks.js:
```diff
@@ -5,29 +5,30 @@
 			}
 
-			const { payment_intent: intentionId } = getCharge( chargeId );
+			const { id: intentionId } = getPaymentIntent( id );
```
Good call! I have removed this logic and the conditional branch on [035eeea](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/035eeea7be8dd3a63a96b219bcf6c285f9283cda).

EduardoUmpierre commented on client/data/payment-intents/hooks.js:
```diff
@@ -0,0 +1,79 @@
+		isLoading,
+		error: getChargeError( chargeId ),
+	};
+};
```
I agree with you, by doing this the hook can focus on just handling the data. The reuse of the existing `getCharge` selector was also a pretty nice catch! Fixed on [035eeea](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/035eeea7be8dd3a63a96b219bcf6c285f9283cda).

EduardoUmpierre commented on client/data/payment-intents/hooks.js:
```diff
@@ -0,0 +1,79 @@
+	};
+};
+
+export const usePaymentIntentFallback = ( id ) =>
```
Fixed on [f6f948a](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/f6f948a03117cd4def980dd4486cded52615470e)!

EduardoUmpierre commented on client/data/payment-intents/test/resolvers.js:
```diff
@@ -0,0 +1,85 @@
+	},
+};
+
+describe( 'getPaymentIntent resolver', () => {
```
Sure, I added tests to the hook and also for the selectors on [ff3a624](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/ff3a624d528fa31cba6f3a23e7bfa8aad04d0f27) and [85e80d1](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/85e80d1a9d2bf32ac23e4667c22c69c5840daf0e).

EduardoUmpierre commented:
Thanks for your thorough review, @luizreis! I have commented on and addressed all the points that you raised in your comments.

Please take a look at it again 🙏

luizreis reviewed (commented):
@EduardoUmpierre - Thanks for all the work addressing the previous comments, this is looking great! I left some small comments, but we're pretty close to merging this PR. 🎉

EduardoUmpierre commented on client/data/timeline/hooks.js:
```diff
@@ -5,43 +5,33 @@
 import { STORE_NAME } from '../constants';
 
-export const useTimeline = ( chargeId ) =>
+export const useTimeline = ( id ) =>
```
Yes, it should always be a payment intent ID. Fixed on [3585daf](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/3585daf0e4b2223e5a0106b3597165d0a08db21f).

EduardoUmpierre commented on client/data/timeline/hooks.js:
```diff
@@ -5,43 +5,33 @@
+					timelineError: error,
+					isLoading: isLoading,
 				};
 			}
```
We're receiving the payment intent ID directly, so we don't really need this payment intent check in this hook. Fixed on [97965fd](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/97965fde3c5b8db8f200dc6c28bec846150c3db0).

EduardoUmpierre commented on client/payment-details/index.js:
```diff
@@ -26,16 +31,36 @@ const PaymentDetails = ( props ) => {
+		error,
+		isLoading: isLoadingData,
+	} = usePaymentIntentWithChargeFallback( id );
+	const isChargeId = useMemo( () => getIsChargeId( id ), [ id ] );
```
My first thought was to add `useMemo` as `isChargeId` is a dependency of the `useEffect`, but we don't really benefit from it in this case. Removed on [6dffd4c](https://github.com/Automattic/woocommerce-payments/pull/4348/commits/6dffd4c7735a125383a96805bea551af081a2ffb).

luizreis commented on package-lock.json:
```diff
@@ -15170,7 +15170,7 @@
       "version": "1.1.1",
       "resolved": "https://registry.npmjs.org/array-flatten/-/array-flatten-1.1.1.tgz",
-      "integrity": "sha1-ml9pkFGx5wczKPKgCJaLZOopVdI=",
+      "integrity": "sha512-PCVAQswWemu6UdxsDFFX/+gVeYqKAod3D3UVm91jHwynguOwAvYPhx8nNlM++NqRcK6CxxpUafjmhIdKiHibqg==",
```
Could these changes to `package-lock.json` cause any issues? I'm approving the PR, but deferring this to you, as it seems a bit strange to change a few integrity hashes to sha512 while leaving some as sha1.

EduardoUmpierre replied:
I wouldn't expect issues from this as there are several references of sha512 and sha1 in this file.

luizreis reviewed (approved):
Thanks for addressing all the review comments @EduardoUmpierre, it's looking great! 🎉 

This PR sets a pretty good foundation to allow us to use payment intent IDs in the links instead of charge IDs and to infer the other statuses that rely on a payment intent object :shipit:
